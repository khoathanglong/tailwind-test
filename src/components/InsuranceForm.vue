<template>
  <div>
    <h1>Tell us about your self</h1>
    <div>
      <label for="Name">Name</label>
      <input
        type="text"
        name="Name"
        placeholder="Add text"
        v-model="name"
        required
      />
      <div v-if="error.name" data-test="nameError">{{ error.name }}</div>
    </div>
    <div>
      <label for="Age">Age</label>
      <input
        type="number"
        name="Age"
        placeholder="Age"
        v-model.number="age"
        required
      />
      <div v-if="error.age" data-test="ageError">{{ error.age }}</div>
    </div>

    <label for="location">Where do you live</label>
    <select name="location" id="location" v-model="selectedLocation">
      <option
        v-for="location in locations"
        :value="location.value"
        :key="location.value"
      >
        {{ location.name }}
      </option>
    </select>
    <div v-for="(pack, index) in packages" :key="pack.value">
      <input
        :id="pack.value"
        type="radio"
        :value="pack.value"
        v-model="selectedPackage"
      />
      <label :for="pack.value">
        {{ pack.name }} {{ packageAdditionalInfo[index] }}</label
      >
    </div>

    <h1 data-test="premium">
      Your premium is {{ premium }}{{ this.currentLocation.currency }}
    </h1>

    <button @click="handleBack" data-test="backButton">Back</button>
    <button @click="handleNext" data-test="nextButton">Next</button>
  </div>
</template>

<script>
export default {
  name: "InsuranceForm",
  data() {
    return {
      name: "",
      age: 20,
      selectedLocation: "hongkong",
      selectedPackage: "standard",
      error: {
        name: "",
        age: "",
      },
      locations: [
        { name: "Hong Kong", currency: "HKD", value: "hongkong", rate: 1 },
        { name: "USA", currency: "USD", value: "usa", rate: 2 },
        { name: "Australia", currency: "AUD", value: "australia", rate: 3 },
      ],
      packages: [
        { value: "standard", name: "Standard", rate: 0 },
        { value: "safe", name: "Safe", rate: 0.5 },
        { value: "extra", name: "Extra", rate: 0.75 },
      ],
    }
  },

  methods: {
    checkFormError() {
      let error = false
      if (!this.name) {
        error = true
        this.error.name = "Name is required"
      }
      if (!this.age) {
        error = true
        this.error.age = "Age is required"
      }
      return error
    },
    handleNext() {
      const hasError = this.checkFormError()
      if (hasError) {
        return
      }

      if (this.age > 100) {
        this.$router.push("/age-error")
      }

      this.$emit("submit", {
        name: this.name,
        age: this.age,
        premium: this.premium,
        currency: this.currentLocation.currency,
        location: this.currentLocation.name,
        pack: this.selectedPackage,
      })
    },
    handleBack() {
      this.$router.push("/")
    },
  },
  computed: {
    currentLocation() {
      return this.locations.find(
        (location) => location.value === this.selectedLocation
      )
    },
    currentPackage() {
      return this.packages.find((pack) => pack.value === this.selectedPackage)
    },
    basePrice() {
      // based on age and location
      return 10 * (this.age || 0) * this.currentLocation.rate
    },
    premium() {
      return this.basePrice * (1 + this.currentPackage.rate)
    },
    packageAdditionalInfo() {
      return this.packages.map((pack) => {
        if (pack.value === "standard") return ""
        return `+(${this.basePrice * pack.rate}${
          this.currentLocation.currency
        }, ${pack.rate * 100}%)`
      })
    },
  },
}
</script>
