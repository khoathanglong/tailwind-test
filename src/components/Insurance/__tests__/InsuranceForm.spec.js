import InsuranceForm from "../InsuranceForm"
import { mount } from "@vue/test-utils"

async function setValue({ wrapper, age, location, pack }) {
  await wrapper.find("input[name=Age]").setValue(age)
  const select = await wrapper.find("select").element
  select.value = location
  select.dispatchEvent(new Event("change"))
  await wrapper.find(`#${pack}`).setChecked()
}

describe("InsuranceForm", () => {
  test("returns to / when click back button", async () => {
    const mockRouter = {
      push: jest.fn(),
    }

    const wrapper = mount(InsuranceForm, {
      mocks: {
        $router: mockRouter,
      },
    })

    await wrapper.find("[data-test=backButton]").trigger("click")
    expect(mockRouter.push).toBeCalledWith("/")
  })

  test("shows name error when click Next button if name is not provided", async () => {
    const wrapper = mount(InsuranceForm)
    await wrapper.find("input[name=Name]").setValue("")
    await wrapper.find("[data-test=nextButton]").trigger("click")
    expect(wrapper.find("[data-test=nameError]").text()).toContain(
      "Name is required"
    )
  })

  test("shows age error when click Next button if age is not provided", async () => {
    const wrapper = mount(InsuranceForm)
    await wrapper.find("input[name=Age]").setValue(null)
    await wrapper.find("[data-test=nextButton]").trigger("click")
    expect(wrapper.find("[data-test=ageError]").text()).toContain(
      "Age is required"
    )
  })

  test("redirects to age-error page if age > 100", async () => {
    const mockRouter = {
      push: jest.fn(),
    }

    const wrapper = mount(InsuranceForm, {
      mocks: {
        $router: mockRouter,
      },
    })
    await wrapper.find("input[name=Age]").setValue(101)
    await wrapper.find("input[name=Name]").setValue("John Doe")
    await wrapper.find("[data-test=nextButton]").trigger("click")

    expect(mockRouter.push).toBeCalledWith("/age-error")
  })

  test("emits submit event if age and name are provided", async () => {
    const mockEmit = jest.fn()
    const wrapper = mount(InsuranceForm, {
      mocks: {
        $emit: mockEmit,
      },
    })
    await wrapper.find("input[name=Name]").setValue("John Doe")
    await setValue({ wrapper, age: 10, location: "hongkong", pack: "standard" })
    await wrapper.find("[data-test=nextButton]").trigger("click")

    expect(mockEmit).toBeCalledWith("submit", {
      age: 10,
      name: "John Doe",
      currency: "HKD",
      pack: "standard",
      location: "Hong Kong",
      premium: 100,
    })
  })

  test("shows correct package option labels based on location", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "hongkong", pack: "standard" })
    expect(wrapper.html()).toMatchSnapshot()
    await setValue({ wrapper, location: "usa", pack: "standard", age: 10 })
    expect(wrapper.html()).toMatchSnapshot()
    await setValue({
      wrapper,
      location: "australia",
      pack: "standard",
      age: 10,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("shows correct premium for Standard Package in Hong Kong", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "hongkong", pack: "standard" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("100HKD")
  })

  test("shows correct premium for Safe Package in Hong Kong", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "hongkong", pack: "safe" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("150HKD")
  })

  test("shows correct premium for Extra Package in Hong Kong", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "hongkong", pack: "extra" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("175HKD")
  })

  test("shows correct premium for Standard Package in USA", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "usa", pack: "standard" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("200USD")
  })

  test("shows correct premium for Safe Package in USA", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "usa", pack: "safe" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("300USD")
  })

  test("shows correct premium for Extra Package in USA", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({ wrapper, age: 10, location: "usa", pack: "extra" })
    expect(wrapper.find("[data-test=premium]").text()).toContain("350USD")
  })

  test("shows correct premium for Standard Package in Australia", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({
      wrapper,
      age: 10,
      location: "australia",
      pack: "standard",
    })
    expect(wrapper.find("[data-test=premium]").text()).toContain("300AUD")
  })

  test("shows correct premium for Safe Package in Australia", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({
      wrapper,
      age: 10,
      location: "australia",
      pack: "safe",
    })
    expect(wrapper.find("[data-test=premium]").text()).toContain("450AUD")
  })

  test("shows correct premium for Extra Package in Australia and renders", async () => {
    const wrapper = mount(InsuranceForm)
    await setValue({
      wrapper,
      age: 10,
      location: "australia",
      pack: "extra",
    })
    expect(wrapper.find("[data-test=premium]").text()).toContain("525AUD")
  })
})
