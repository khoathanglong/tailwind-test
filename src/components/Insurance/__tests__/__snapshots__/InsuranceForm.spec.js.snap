// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsuranceForm shows correct package option labels based on location 1`] = `
<div class="h-screen w-fit flex flex-col justify-center">
  <h1 class="text-3xl font-bold self-center">Tell us about your self</h1>
  <div class="mt-10 self-center"><label for="Name" class="block">Name</label> <input type="text" name="Name" placeholder="Add text" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="nameError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="Age" class="block">Age</label> <input type="number" name="Age" placeholder="Age" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="ageError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="location" class="block">Where do you live</label> <select name="location" id="location" class="rounded w-60 h-10 border-2 border-grey-200 bg-white">
      <option value="hongkong">
        Hong Kong
      </option>
      <option value="usa">
        USA
      </option>
      <option value="australia">
        Australia
      </option>
    </select></div>
  <div class="self-center">
    <div class="self-start mt-5 w-60"><input id="standard" type="radio" value="standard"> <label for="standard">
        Standard </label></div>
    <div class="self-start mt-5 w-60"><input id="safe" type="radio" value="safe"> <label for="safe">
        Safe +(50HKD, 50%)</label></div>
    <div class="self-start mt-5 w-60"><input id="extra" type="radio" value="extra"> <label for="extra">
        Extra +(75HKD, 75%)</label></div>
  </div>
  <h1 data-test="premium" class="text-2xl font-bold self-center my-10">
    Your premium is 100HKD
  </h1>
  <div class="self-center"><button data-test="backButton" class="rounded w-24 h-10 border-2 border-grey-900 text-black mx-2">
      Back
    </button> <button data-test="nextButton" class="rounded w-24 h-10 bg-black text-white">
      Next
    </button></div>
</div>
`;

exports[`InsuranceForm shows correct package option labels based on location 2`] = `
<div class="h-screen w-fit flex flex-col justify-center">
  <h1 class="text-3xl font-bold self-center">Tell us about your self</h1>
  <div class="mt-10 self-center"><label for="Name" class="block">Name</label> <input type="text" name="Name" placeholder="Add text" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="nameError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="Age" class="block">Age</label> <input type="number" name="Age" placeholder="Age" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="ageError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="location" class="block">Where do you live</label> <select name="location" id="location" class="rounded w-60 h-10 border-2 border-grey-200 bg-white">
      <option value="hongkong">
        Hong Kong
      </option>
      <option value="usa">
        USA
      </option>
      <option value="australia">
        Australia
      </option>
    </select></div>
  <div class="self-center">
    <div class="self-start mt-5 w-60"><input id="standard" type="radio" value="standard"> <label for="standard">
        Standard </label></div>
    <div class="self-start mt-5 w-60"><input id="safe" type="radio" value="safe"> <label for="safe">
        Safe +(100USD, 50%)</label></div>
    <div class="self-start mt-5 w-60"><input id="extra" type="radio" value="extra"> <label for="extra">
        Extra +(150USD, 75%)</label></div>
  </div>
  <h1 data-test="premium" class="text-2xl font-bold self-center my-10">
    Your premium is 200USD
  </h1>
  <div class="self-center"><button data-test="backButton" class="rounded w-24 h-10 border-2 border-grey-900 text-black mx-2">
      Back
    </button> <button data-test="nextButton" class="rounded w-24 h-10 bg-black text-white">
      Next
    </button></div>
</div>
`;

exports[`InsuranceForm shows correct package option labels based on location 3`] = `
<div class="h-screen w-fit flex flex-col justify-center">
  <h1 class="text-3xl font-bold self-center">Tell us about your self</h1>
  <div class="mt-10 self-center"><label for="Name" class="block">Name</label> <input type="text" name="Name" placeholder="Add text" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="nameError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="Age" class="block">Age</label> <input type="number" name="Age" placeholder="Age" required="required" class="rounded w-60 py-2 h-10 border-2 border-grey-200">
    <div data-test="ageError" class="text-rose-500">

    </div>
  </div>
  <div class="self-center mt-5"><label for="location" class="block">Where do you live</label> <select name="location" id="location" class="rounded w-60 h-10 border-2 border-grey-200 bg-white">
      <option value="hongkong">
        Hong Kong
      </option>
      <option value="usa">
        USA
      </option>
      <option value="australia">
        Australia
      </option>
    </select></div>
  <div class="self-center">
    <div class="self-start mt-5 w-60"><input id="standard" type="radio" value="standard"> <label for="standard">
        Standard </label></div>
    <div class="self-start mt-5 w-60"><input id="safe" type="radio" value="safe"> <label for="safe">
        Safe +(150AUD, 50%)</label></div>
    <div class="self-start mt-5 w-60"><input id="extra" type="radio" value="extra"> <label for="extra">
        Extra +(225AUD, 75%)</label></div>
  </div>
  <h1 data-test="premium" class="text-2xl font-bold self-center my-10">
    Your premium is 300AUD
  </h1>
  <div class="self-center"><button data-test="backButton" class="rounded w-24 h-10 border-2 border-grey-900 text-black mx-2">
      Back
    </button> <button data-test="nextButton" class="rounded w-24 h-10 bg-black text-white">
      Next
    </button></div>
</div>
`;
